File: scripts/multimodal_browser_smoke.py
#!/usr/bin/env python3
"""Real-model browser smoke for multimodal bridge prompt ingestion."""

from __future__ import annotations

import argparse
import asyncio
import json
import os
from pathlib import Path
import shutil
import sys
import tempfile

from state_persistence_browser_smoke import (
    copy_artifacts,
    download_model,
    require,
    run_playwright,
    serve,
    validate_model_hash,
)

DEFAULT_MODEL_CACHE = (
    Path.home() / ".cache" / "llama-web-bridge" / "multimodal-smoke-models"
)


def resolve_file(
    *,
    path: Path | None,
    url: str,
    expected_sha256: str,
    cache_dir: Path,
    label: str,
) -> Path:
    require(bool(expected_sha256), f"{label} SHA-256 is required")
    if path is not None:
        resolved = path.expanduser().resolve()
        require(resolved.is_file(), f"{label} path does not exist: {resolved}")
        validate_model_hash(resolved, expected_sha256)
        return resolved
    require(bool(url), f"{label} URL or local path is required")
    return download_model(
        url,
        cache_dir.expanduser().resolve(),
        expected_sha256,
    )


def write_harness(web_root: Path) -> None:
    script = """
<!doctype html>
<meta charset="utf-8">
<title>llama-web-bridge multimodal smoke</title>
<pre id="result">pending</pre>
<script type="module">
(async () => {
  const result = document.getElementById('result');
  const finish = (payload) => {
    result.textContent = JSON.stringify(payload);
    window.__smokeResult = payload;
  };
  const assert = (condition, message) => {
    if (!condition) {
      throw new Error(message);
    }
  };
  const createImageBytes = async () => {
    const canvas = document.createElement('canvas');
    canvas.width = 320;
    canvas.height = 180;
    const context = canvas.getContext('2d');
    context.fillStyle = '#0f172a';
    context.fillRect(0, 0, canvas.width, canvas.height);
    context.fillStyle = '#22d3ee';
    context.fillRect(16, 16, 288, 148);
    context.fillStyle = '#111827';
    context.font = 'bold 42px sans-serif';
    context.fillText('HELLO', 80, 108);
    const blob = await new Promise((resolve, reject) => {
      canvas.toBlob((value) => {
        if (value) {
          resolve(value);
        } else {
          reject(new Error('synthetic image encoding failed'));
        }
      }, 'image/png');
    });
    return new Uint8Array(await blob.arrayBuffer());
  };

  try {
    assert(window.crossOriginIsolated, 'test page is not cross-origin isolated');
    const module = await import('/llama_webgpu_bridge.js');
    const LlamaWebGpuBridge =
      module.LlamaWebGpuBridge || window.LlamaWebGpuBridge;
    assert(
      typeof LlamaWebGpuBridge === 'function',
      'LlamaWebGpuBridge export was not registered',
    );

    const imageBytes = await createImageBytes();
    const modeResults = [];
    const verifyMode = async (disableWorker, mode) => {
      const bridge = new LlamaWebGpuBridge({ disableWorker, logLevel: 1 });
      const startedAt = performance.now();
      try {
        await bridge.loadModelFromUrl('/multimodal-model.gguf', {
          nCtx: 4096,
          nGpuLayers: 0,
          nThreads: 4,
          nBatch: 512,
          nUbatch: 256,
          useCache: false,
          forceRemoteFetchBackend: false,
        });
        await bridge.loadMultimodalProjector('/multimodal-mmproj.gguf');
        assert(bridge.supportsVision(), `${mode} did not report vision support`);

        const output = await bridge.createCompletion('what do you see?', {
          nPredict: 64,
          temp: 0,
          topK: 1,
          topP: 1,
          seed: 42,
          tokenEventEncoding: 'text',
          parts: [{ type: 'image', bytes: imageBytes }],
          mediaMaxImagePixels: 307200,
          mediaMaxImageEdge: 768,
        });
        const outputText = String(output || '').trim();
        assert(outputText.length > 0, `${mode} returned empty multimodal output`);
        assert(
          outputText.toLowerCase().includes('hello'),
          `${mode} did not recognize the synthetic HELLO image: ${outputText}`,
        );
        modeResults.push({
          mode,
          elapsedMs: Math.round(performance.now() - startedAt),
          output: outputText.slice(0, 160),
        });
      } finally {
        await bridge.dispose();
      }
    };

    await verifyMode(true, 'direct runtime');
    await verifyMode(false, 'worker runtime');
    finish({
      ok: true,
      modes: ['direct runtime', 'worker runtime'],
      modeResults,
    });
  } catch (error) {
    finish({
      ok: false,
      error: String(error && error.stack ? error.stack : error),
    });
  }
})();
</script>
"""
    (web_root / "index.html").write_text(script, encoding="utf-8")


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument(
        "--dist-dir",
        type=Path,
        default=Path(os.environ.get("BRIDGE_DIST_DIR", "dist")),
        help="Directory containing built bridge artifacts.",
    )
    parser.add_argument(
        "--timeout-ms",
        type=int,
        default=int(os.environ.get("LLAMA_WEBGPU_MULTIMODAL_TIMEOUT_MS", "420000")),
        help="Browser operation timeout in milliseconds.",
    )
    parser.add_argument(
        "--model-url",
        default=os.environ.get("LLAMA_WEBGPU_MULTIMODAL_MODEL_URL", ""),
        help="Qwen multimodal GGUF URL.",
    )
    parser.add_argument(
        "--model-path",
        type=Path,
        default=(
            Path(os.environ["LLAMA_WEBGPU_MULTIMODAL_MODEL_PATH"])
            if os.environ.get("LLAMA_WEBGPU_MULTIMODAL_MODEL_PATH")
            else None
        ),
        help="Local Qwen multimodal GGUF path.",
    )
    parser.add_argument(
        "--model-sha256",
        default=os.environ.get("LLAMA_WEBGPU_MULTIMODAL_MODEL_SHA256", ""),
        help="Expected model SHA-256.",
    )
    parser.add_argument(
        "--mmproj-url",
        default=os.environ.get("LLAMA_WEBGPU_MULTIMODAL_MMPROJ_URL", ""),
        help="Qwen multimodal projector GGUF URL.",
    )
    parser.add_argument(
        "--mmproj-path",
        type=Path,
        default=(
            Path(os.environ["LLAMA_WEBGPU_MULTIMODAL_MMPROJ_PATH"])
            if os.environ.get("LLAMA_WEBGPU_MULTIMODAL_MMPROJ_PATH")
            else None
        ),
        help="Local Qwen multimodal projector GGUF path.",
    )
    parser.add_argument(
        "--mmproj-sha256",
        default=os.environ.get("LLAMA_WEBGPU_MULTIMODAL_MMPROJ_SHA256", ""),
        help="Expected multimodal projector SHA-256.",
    )
    parser.add_argument(
        "--model-cache-dir",
        type=Path,
        default=Path(
            os.environ.get(
                "LLAMA_WEBGPU_MULTIMODAL_MODEL_CACHE",
                DEFAULT_MODEL_CACHE,
            )
        ),
        help="Cache directory used for downloaded model files.",
    )
    parser.add_argument(
        "--artifacts-dir",
        type=Path,
        default=(
            Path(os.environ["LLAMA_WEBGPU_MULTIMODAL_ARTIFACTS_DIR"])
            if os.environ.get("LLAMA_WEBGPU_MULTIMODAL_ARTIFACTS_DIR")
            else None
        ),
        help="Directory for JSON, console, and screenshot diagnostics.",
    )
    return parser.parse_args()


def main() -> int:
    args = parse_args()
    dist_dir = args.dist_dir.resolve()
    require(dist_dir.is_dir(), f"dist directory does not exist: {dist_dir}")
    cache_dir = args.model_cache_dir.expanduser().resolve()
    model_path = resolve_file(
        path=args.model_path,
        url=args.model_url,
        expected_sha256=args.model_sha256,
        cache_dir=cache_dir,
        label="multimodal model",
    )
    mmproj_path = resolve_file(
        path=args.mmproj_path,
        url=args.mmproj_url,
        expected_sha256=args.mmproj_sha256,
        cache_dir=cache_dir,
        label="multimodal projector",
    )
    artifacts_dir = args.artifacts_dir.resolve() if args.artifacts_dir else None

    with tempfile.TemporaryDirectory(prefix="llama-web-bridge-multimodal-") as tmp:
        web_root = Path(tmp)
        copy_artifacts(dist_dir, web_root)
        shutil.copyfile(model_path, web_root / "multimodal-model.gguf")
        shutil.copyfile(mmproj_path, web_root / "multimodal-mmproj.gguf")
        write_harness(web_root)
        if artifacts_dir is not None:
            artifacts_dir.mkdir(parents=True, exist_ok=True)
            shutil.copyfile(web_root / "index.html", artifacts_dir / "index.html")
        with serve(web_root) as url:
            payload = asyncio.run(
                run_playwright(
                    url,
                    args.timeout_ms,
                    artifacts_dir,
                    artifact_prefix="multimodal-smoke",
                )
            )

    print(json.dumps(payload, indent=2, sort_keys=True))
    if payload.get("ok") is not True:
        return 1
    require(
        payload.get("modes") == ["direct runtime", "worker runtime"],
        "multimodal smoke modes payload mismatch",
    )
    mode_results = payload.get("modeResults")
    require(
        isinstance(mode_results, list) and len(mode_results) == 2,
        "multimodal mode results missing",
    )
    return 0


if __name__ == "__main__":
    try:
        raise SystemExit(main())
    except Exception as exc:  # noqa: BLE001 - top-level CLI guard
        print(f"multimodal browser smoke failed: {exc}", file=sys.stderr)
        raise SystemExit(1)

File: scripts/multimodal_browser_smoke.mjs
#!/usr/bin/env node
// Real-model browser smoke for multimodal bridge prompt ingestion.
//
// Ported from multimodal_browser_smoke.py with the same flags, environment
// variables, harness page and output.

import fsp from 'node:fs/promises';
import path from 'node:path';
import process from 'node:process';

import {
  copyBridgeArtifacts,
  ensure,
  env,
  expandHome,
  isDirectory,
  parseSmokeArgs,
  pyEquals,
  pyGet,
  pyJson,
  resolvePath,
  resolvePinnedFile,
  runMain,
  runPlaywright,
  withServer,
  withTempDir,
  writeStdout,
} from './browser_smoke_support.mjs';

const DESCRIPTION = 'Real-model browser smoke for multimodal bridge prompt ingestion.';
const DEFAULT_MODEL_CACHE = '~/.cache/llama-web-bridge/multimodal-smoke-models';
const MODES = Object.freeze(['direct runtime', 'worker runtime']);

export function renderHarness() {
  return `
<!doctype html>
<meta charset="utf-8">
<title>llama-web-bridge multimodal smoke</title>
<pre id="result">pending</pre>
<script type="module">
(async () => {
  const result = document.getElementById('result');
  const finish = (payload) => {
    result.textContent = JSON.stringify(payload);
    window.__smokeResult = payload;
  };
  const assert = (condition, message) => {
    if (!condition) {
      throw new Error(message);
    }
  };
  const createImageBytes = async () => {
    const canvas = document.createElement('canvas');
    canvas.width = 320;
    canvas.height = 180;
    const context = canvas.getContext('2d');
    context.fillStyle = '#0f172a';
    context.fillRect(0, 0, canvas.width, canvas.height);
    context.fillStyle = '#22d3ee';
    context.fillRect(16, 16, 288, 148);
    context.fillStyle = '#111827';
    context.font = 'bold 42px sans-serif';
    context.fillText('HELLO', 80, 108);
    const blob = await new Promise((resolve, reject) => {
      canvas.toBlob((value) => {
        if (value) {
          resolve(value);
        } else {
          reject(new Error('synthetic image encoding failed'));
        }
      }, 'image/png');
    });
    return new Uint8Array(await blob.arrayBuffer());
  };

  try {
    assert(window.crossOriginIsolated, 'test page is not cross-origin isolated');
    const module = await import('/llama_webgpu_bridge.js');
    const LlamaWebGpuBridge =
      module.LlamaWebGpuBridge || window.LlamaWebGpuBridge;
    assert(
      typeof LlamaWebGpuBridge === 'function',
      'LlamaWebGpuBridge export was not registered',
    );

    const imageBytes = await createImageBytes();
    const modeResults = [];
    const verifyMode = async (disableWorker, mode) => {
      const bridge = new LlamaWebGpuBridge({ disableWorker, logLevel: 1 });
      const startedAt = performance.now();
      try {
        await bridge.loadModelFromUrl('/multimodal-model.gguf', {
          nCtx: 4096,
          nGpuLayers: 0,
          nThreads: 4,
          nBatch: 512,
          nUbatch: 256,
          useCache: false,
          forceRemoteFetchBackend: false,
        });
        await bridge.loadMultimodalProjector('/multimodal-mmproj.gguf');
        assert(bridge.supportsVision(), \`\${mode} did not report vision support\`);

        // Normal image ingestion resizes through RGB. Also exercise the native
        // file/encoded entry points so helper API compatibility cannot hide
        // behind that independent path. The direct runtime owns this FS.
        if (disableWorker) {
          const core = bridge._runtime?._core;
          assert(core?.FS && typeof core.ccall === 'function',
            'direct runtime core is unavailable for media compatibility checks');
          const path = '/mtmd-compat.png';
          core.FS.writeFile(path, imageBytes);
          try {
            assert(
              core.ccall('llamadart_webgpu_media_add_file', 'number',
                ['string'], [path]) === 0,
              'native file helper failed to decode the real PNG',
            );
            core.ccall('llamadart_webgpu_media_clear_pending', null, [], []);
            assert(
              core.ccall('llamadart_webgpu_media_add_encoded', 'number',
                ['array', 'number'], [imageBytes, imageBytes.length]) === 0,
              'native buffer helper failed to decode the real PNG',
            );
            core.ccall('llamadart_webgpu_media_clear_pending', null, [], []);
            assert(
              core.ccall('llamadart_webgpu_media_add_encoded', 'number',
                ['array', 'number'], [imageBytes, 0]) === -3,
              'native buffer helper must reject empty input',
            );
            assert(
              core.ccall('llamadart_webgpu_media_add_encoded', 'number',
                ['array', 'number'], [imageBytes.subarray(0, 8), 8]) === -4,
              'native buffer helper must reject a truncated PNG',
            );
            assert(
              core.ccall('llamadart_webgpu_media_add_file', 'number',
                ['string'], ['/mtmd-compat-missing.png']) === -4,
              'native file helper must reject a missing file',
            );
          } finally {
            core.ccall('llamadart_webgpu_media_clear_pending', null, [], []);
            core.FS.unlink(path);
          }
        }

        const usages = [];
        const promptTextTokens = (await bridge.tokenize('what do you see?', true)).length;
        const output = await bridge.createCompletion('what do you see?', {
          onUsage: (usage) => usages.push(usage),
          nPredict: 64,
          temp: 0,
          topK: 1,
          topP: 1,
          seed: 42,
          tokenEventEncoding: 'text',
          parts: [{ type: 'image', bytes: imageBytes }],
          mediaMaxImagePixels: 50000,
          mediaMaxImageEdge: 300,
        });
        const outputText = String(output || '').trim();
        assert(outputText.length > 0, \`\${mode} returned empty multimodal output\`);
        assert(
          outputText.toLowerCase().includes('hello'),
          \`\${mode} did not recognize the synthetic HELLO image: \${outputText}\`,
        );
        assert(usages.length === 1, \`\${mode} reported usage \${usages.length} times\`);
        const [usage] = usages;
        assert(
          usage.promptTokens > promptTextTokens,
          \`\${mode} reported \${usage.promptTokens} prompt positions for an image prompt of \${promptTextTokens} text tokens\`,
        );
        assert(usage.cachedPromptTokens === 0, \`\${mode} reused \${usage.cachedPromptTokens} tokens of a multimodal prompt\`);
        assert(
          usage.completionTokens > 0 && usage.completionTokens <= 64,
          \`\${mode} reported \${usage.completionTokens} completion tokens\`,
        );
        assert(
          usage.finishReason === (usage.completionTokens === 64 ? 'length' : 'stop'),
          \`\${mode} finished with \${usage.finishReason} after \${usage.completionTokens} tokens\`,
        );
        assert(
          usage.timeToFirstTokenMs > 0 && usage.timeToFirstTokenMs <= usage.durationMs,
          \`\${mode} reported time to first token \${usage.timeToFirstTokenMs} of \${usage.durationMs} ms\`,
        );
        const metadata = bridge.getModelMetadata();
        const runtimeNotes = String(metadata['llamadart.webgpu.runtime_notes'] || '');
        assert(
          runtimeNotes.includes('media_image_resized:320x180->'),
          \`\${mode} did not expose the image-resize diagnostic: \${runtimeNotes}\`,
        );
        modeResults.push({
          mode,
          elapsedMs: Math.round(performance.now() - startedAt),
          output: outputText.slice(0, 160),
          promptTextTokens,
          usage,
          imageResizeDiagnostic: runtimeNotes
            .split(';')
            .find((note) => note.startsWith('media_image_resized:')),
        });
      } finally {
        await bridge.dispose();
      }
    };

    await verifyMode(true, 'direct runtime');
    await verifyMode(false, 'worker runtime');
    finish({
      ok: true,
      modes: ['direct runtime', 'worker runtime'],
      modeResults,
    });
  } catch (error) {
    finish({
      ok: false,
      error: String(error && error.stack ? error.stack : error),
    });
  }
})();
</script>
`;
}

export function parseArgs(argv) {
  return parseSmokeArgs(argv, {
    prog: 'multimodal_browser_smoke.mjs',
    description: DESCRIPTION,
    options: [
      {
        flag: '--dist-dir',
        type: 'path',
        default: () => env.path('BRIDGE_DIST_DIR', 'dist'),
        help: 'Directory containing built bridge artifacts.',
      },
      {
        flag: '--timeout-ms',
        type: 'int',
        default: () => env.int('LLAMA_WEBGPU_MULTIMODAL_TIMEOUT_MS', '420000'),
        help: 'Browser operation timeout in milliseconds.',
      },
      {
        flag: '--model-url',
        type: 'string',
        default: () => env.string('LLAMA_WEBGPU_MULTIMODAL_MODEL_URL'),
        help: 'Qwen multimodal GGUF URL.',
      },
      {
        flag: '--model-path',
        type: 'path',
        default: () => env.optionalPath('LLAMA_WEBGPU_MULTIMODAL_MODEL_PATH'),
        help: 'Local Qwen multimodal GGUF path.',
      },
      {
        flag: '--model-sha256',
        type: 'string',
        default: () => env.string('LLAMA_WEBGPU_MULTIMODAL_MODEL_SHA256'),
        help: 'Expected model SHA-256.',
      },
      {
        flag: '--mmproj-url',
        type: 'string',
        default: () => env.string('LLAMA_WEBGPU_MULTIMODAL_MMPROJ_URL'),
        help: 'Qwen multimodal projector GGUF URL.',
      },
      {
        flag: '--mmproj-path',
        type: 'path',
        default: () => env.optionalPath('LLAMA_WEBGPU_MULTIMODAL_MMPROJ_PATH'),
        help: 'Local Qwen multimodal projector GGUF path.',
      },
      {
        flag: '--mmproj-sha256',
        type: 'string',
        default: () => env.string('LLAMA_WEBGPU_MULTIMODAL_MMPROJ_SHA256'),
        help: 'Expected multimodal projector SHA-256.',
      },
      {
        flag: '--model-cache-dir',
        type: 'path',
        default: () => env.path('LLAMA_WEBGPU_MULTIMODAL_MODEL_CACHE', DEFAULT_MODEL_CACHE),
        help: 'Cache directory used for downloaded model files.',
      },
      {
        flag: '--artifacts-dir',
        type: 'path',
        default: () => env.optionalPath('LLAMA_WEBGPU_MULTIMODAL_ARTIFACTS_DIR'),
        help: 'Directory for JSON, console, and screenshot diagnostics.',
      },
    ],
  });
}

// Checks that run after the payload is printed.
export function checkPayload(payload) {
  ensure(pyEquals(pyGet(payload, 'modes'), MODES), 'multimodal smoke modes payload mismatch');
  const modeResults = pyGet(payload, 'modeResults');
  ensure(Array.isArray(modeResults) && modeResults.length === 2, 'multimodal mode results missing');
}

export async function main(argv = process.argv.slice(2)) {
  const args = parseArgs(argv);
  const distDir = resolvePath(args.distDir);
  ensure(isDirectory(distDir), `dist directory does not exist: ${distDir}`);
  const cacheDir = resolvePath(expandHome(args.modelCacheDir));
  const modelPath = await resolvePinnedFile({
    filePath: args.modelPath,
    url: args.modelUrl,
    expectedSha256: args.modelSha256,
    cacheDir,
    label: 'multimodal model',
  });
  const mmprojPath = await resolvePinnedFile({
    filePath: args.mmprojPath,
    url: args.mmprojUrl,
    expectedSha256: args.mmprojSha256,
    cacheDir,
    label: 'multimodal projector',
  });
  const artifactsDir = args.artifactsDir !== null ? resolvePath(args.artifactsDir) : null;

  const payload = await withTempDir('llama-web-bridge-multimodal-', async (webRoot) => {
    await copyBridgeArtifacts(distDir, webRoot);
    await fsp.copyFile(modelPath, path.join(webRoot, 'multimodal-model.gguf'));
    await fsp.copyFile(mmprojPath, path.join(webRoot, 'multimodal-mmproj.gguf'));
    await fsp.writeFile(path.join(webRoot, 'index.html'), renderHarness(), 'utf8');
    if (artifactsDir !== null) {
      await fsp.mkdir(artifactsDir, { recursive: true });
      await fsp.copyFile(path.join(webRoot, 'index.html'), path.join(artifactsDir, 'index.html'));
    }
    return withServer(webRoot, (url) => runPlaywright(url, args.timeoutMs, artifactsDir, 'multimodal-smoke'));
  });

  writeStdout(`${pyJson(payload, { indent: 2, sortKeys: true })}\n`);
  if (payload.ok !== true) return 1;
  checkPayload(payload);
  return 0;
}

if (import.meta.main) {
  await runMain('multimodal', main);
}
